### PR TITLE
Refactor MetadataCache for Raft metadata

### PR DESCRIPTION
**Pull request by mumrah**
For 2.8 we will keep the existing MetadataCache implementation intact. Instead of replacing it with the changes needed for Raft, we will extract an interface for both implementations to use. KafkaServer (ZK-backed clusters) will only pick the ZK metadata cache impl.

The only real change to production code here will be the use of MetadataBroker instead of Broker for MetadataCache#getAliveBroker and MetadataCache#getAliveBrokers. This was done to eliminate unnecessary divergence between the two implementations.

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala]: Are we just not testing topic Ids here anymore?

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/MetadataCache.scala]: Is there a place where we are setting the topic ID in the metadata response?

**Review by jolshan — COMMENTED** [review]
I'm not sure if we are still supporting topic IDs here.

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (jolshan) [reply core/src/main/scala/kafka/server/MetadataCache.scala]: Let me know if I missed it somewhere, but we will need the ID included in the response.

I think we might need a MetadataImage method like `topicIdToName` but instead `topicNameToId`

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:487]: Why remove this?

Comment (mumrah) [reply core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:487]: Hmm, I think this was an artifact of the merge, I'll restore this test

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (mumrah) [inline core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala]: Since we're looking up the cluster by listener name here, we don't see the offline broker in the MetadataImage because it's endpoints map is empty. This leads to `cluster.leaderFor` on L534 returning null

@hachikuji @cmccabe is a change in metadata behavior, or does this test have bad assumptions

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (cmccabe) [inline core/src/main/scala/kafka/server/MetadataCache.scala:57]: The formatting is kind of weird here.  Why the blank line after `getTopicMetadata(`?

Also, it would be good to have Javadoc for this function

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (cmccabe) [inline core/src/main/scala/kafka/server/MetadataCache.scala:84]: can you put this in javadoc format ?  i.e. `@returns`

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (cmccabe) [inline core/src/main/scala/kafka/server/MetadataCache.scala:118]: This is kind of a lot of code so why not have it in its own file?

The JavaDoc should also explain that this is for brokers using ZK and not for brokers using the metadata quorum.

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (cmccabe) [inline core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:45]: we should have some way of running these tests on the raft metadata cache as well as the zk metadata cache.  I guess we can do that in a follow-up PR, though

Comment (mumrah) [reply core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:45]: Yea, I think we can probably do some parameterization thing or possibly even use test templating similar to #9986

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/MetadataCache.scala:105]: nit: I don't think these factories are providing much over the constructors `MetdataCache.zkMetadataCache` vs `new ZkMetadataCache`

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (mumrah) [reply core/src/main/scala/kafka/server/MetadataCache.scala:118]: I left the ZK implementation in place since it's really the only production implementation for now. It also reduces the size of the diff for this change. I don't feel very strongly about it either way, so I'm happy to relocate it to a separate file

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (mumrah) [reply core/src/main/scala/kafka/server/MetadataCache.scala:84]: Yea, this just got pulled up from the class when I extracted the trait. I'll fix up these comments

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/server/MetadataCache.scala:118]: Perhaps we can do this as a follow-up? It is nice at the moment to see the diffs clearly.

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (mumrah) [reply core/src/main/scala/kafka/server/MetadataCache.scala:105]: I was thinking it might be easier to refactor in the future if we only need to rename the factory method rather than changing all the `new Class`.

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:45]: Should be easy to do with all the fancy junit 5 stuff. Since there are no separate tests for `RaftMetadataCache` (unless I'm missing some), I think we should do it here.

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/server/MetadataCache.scala:105]: Ok. Probably not a ton of work either way, but will leave it up to you.

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/MetadataCache.scala:63]: nit: We can save it for a follow-up, but it would be nice to drop all the `get` prefixes here

File: core/src/main/scala/kafka/server/metadata/MetadataImage.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/metadata/MetadataImage.scala]: nit: remove newline

File: core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala]: nit: unneeded braces

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (mumrah) [reply core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:45]: Yea easy enough. I added `@ParameterizedTest` to MetadataCacheTest. One is failing with the Raft metadata cache, so I left that for ZK-only now.

File: core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala:305]: We shouldn't need this for the RaftMetadataCache. Did we end up rewriting the code to update from the image builder?

File: core/src/main/scala/kafka/server/metadata/MetadataBrokers.scala
Comment (mumrah) [inline core/src/main/scala/kafka/server/metadata/MetadataBrokers.scala:47]: I wanted to go ahead and conform to MetadataBroker for both implementations. One side-effect is we are exposing the fenced flag to ZK-based clusters. I've set it to false here since we don't have any notion of broker fencing in the ZK-based metadata. Is there any harm in this?

File: core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala
Comment (mumrah) [inline core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala]: This test is failing in Raft mode. I'll investigate

File: core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala:305]: Discussed offline. This is useful temporarily to allow reuse of `MetadataCacheTest`. Once 2.8 is cut, we'll hopefully be able to consolidate this logic to use the MetadataImage for both zk and raft.

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (mumrah) [reply core/src/main/scala/kafka/server/MetadataCache.scala:63]: KAFKA-12299

File: core/src/main/scala/kafka/server/MetadataCache.scala
Comment (mumrah) [reply core/src/main/scala/kafka/server/MetadataCache.scala:118]: KAFKA-12299

File: core/src/main/scala/kafka/server/metadata/MetadataBrokers.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/server/metadata/MetadataBrokers.scala:47]: as long as this is just used by the ZK code path there is no harm

**Review by cmccabe — APPROVED** [review]
LGTM

File: core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala
Comment (mumrah) [inline core/src/main/scala/kafka/server/metadata/RaftMetadataCache.scala:315]: Note that here we are, somewhat incorrectly, excluding the broker's rack from the Node object. This was done to keep the behavior compatible with the existing MetadataCache (now ZkMetadataCache). Once we have this merged and 2.8 is released, we can work on determining if this was done for a good reason, or if it's a bug in the existing MetadataCache logic. Since this class isn't used for any production code yet, we can leave this as-is for now.

**Pull request merged**
